### PR TITLE
[INVE-19533] Brush Select - Add support for combo selection

### PR DESCRIPTION
**Pull request by daniele-pini**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Description of change

This PR adds support for selecting combos using the "brush-select" behavior. A new "includeCombos" parameter is added (default = `false`) to explicitly enable this effect.

Combos are selected when their midpoint is included in the rectangular brush area (similarly to edges) - most obvious when the combo is expanded:

https://user-images.githubusercontent.com/16387428/193299690-7177e9cc-7ed5-40a7-aa63-d7dc4f9d9030.mp4

### CODE REVIEW

**Comment by daniele-pini** [comment]
Related G6 PR: https://github.com/antvis/G6/pull/3972

File: packages/pc/package.json [hunk 1]
```diff
@@ -124,4 +124,4 @@
     "webpack-cli": "^3.3.10",
     "worker-loader": "^3.0.0"
   }
-}
+}
```
Comment (Verdagio) [inline packages/pc/package.json:127]: newline

File: packages/element/package.json [hunk 2]
```diff
@@ -87,6 +87,6 @@
     "ts-jest": "^24.1.0",
     "ts-loader": "^7.0.3",
     "typescript": "^4.6.3",
-    "@antv/g6": "4.7.4-siren.0"
+    "@antv/g6": "4.7.4-siren.1"
   }
-}
+}
```
Comment (Verdagio) [inline packages/element/package.json:92]: newline

File: packages/g6/package.json [hunk 3]
```diff
@@ -90,4 +90,4 @@
     "webpack": "^4.41.4",
     "webpack-cli": "^3.3.10"
   }
-}
+}
```
Comment (Verdagio) [inline packages/g6/package.json:93]: newline

File: packages/plugin/package.json [hunk 4]
```diff
@@ -59,6 +59,6 @@
     "jquery": "^3.5.1",
     "rimraf": "^3.0.2",
     "ts-jest": "^26.4.4",
-    "@antv/g6": "4.7.4-siren.0"
+    "@antv/g6": "4.7.4-siren.1"
   }
-}
+}
```
Comment (Verdagio) [inline packages/plugin/package.json:64]: newline

File: packages/react-node/package.json [hunk 5]
```diff
@@ -44,4 +44,4 @@
     "prettier": "^1.19.1",
     "yorkie": "^2.0.0"
   }
-}
+}
```
Comment (Verdagio) [inline packages/react-node/package.json:47]: newline

File: packages/site/package.json [hunk 6]
```diff
@@ -55,4 +55,4 @@
   "resolutions": {
     "@types/react": "^16.9.35"
   }
-}
+}
```
Comment (Verdagio) [inline packages/site/package.json:58]: newline

**Review by Verdagio — CHANGES_REQUESTED** [review]

File: packages/pc/src/behavior/brush-select.ts [hunk 7]
```diff
@@ -170,16 +177,39 @@ export default {
...
+    if (this.includeCombos) {
+      graph.getCombos().forEach((combo) => {
+        if (!combo.isVisible()) return;
+        const bbox = combo.getBBox();
+        if (
+          bbox.centerX >= left &&
+          bbox.centerX <= right &&
+          bbox.centerY >= top &&
+          bbox.centerY <= bottom
+        ) {
```
Comment (Verdagio) [inline packages/pc/src/behavior/brush-select.ts:190]: I wonder how often we have to do an operation like this? It's not very nice to read.... If there are multiple areas where we do this maybe we could make a util function `bboxInbounds(centerX, centerY) => bool` ?

File: packages/pc/src/behavior/brush-select.ts [hunk 8]
```diff
@@ -170,16 +177,39 @@ export default {
...
+    this.selectedCombos = selectedCombos;
     if (this.onSelect) {
-      this.onSelect(selectedNodes, selectedEdges);
+      this.onSelect(selectedNodes, selectedEdges, selectedCombos);
     }
     graph.emit('nodeselectchange', {
       selectedItems: {
         nodes: selectedNodes,
         edges: selectedEdges,
+        combos: selectedCombos,
```
Comment (Verdagio) [inline packages/pc/src/behavior/brush-select.ts:212]:
```suggestion
        combos: selectedCombos
```

File: packages/site/docs/manual/middle/states/defaultBehavior.en.md [hunk 9]
```diff
@@ -519,7 +520,8 @@ const graph = new G6.Graph({
       {
         type: 'brush-select',
         trigger: 'ctrl',
-        includeEdges: false,
+        includeEdges: false, // Will ignore edges
+        includeCombos: true, // Will select combos
```
Comment (Verdagio) [inline packages/site/docs/manual/middle/states/defaultBehavior.en.md:524]:
```suggestion
        includeCombos: true // Will select combos
```

File: packages/pc/src/behavior/brush-select.ts [hunk 10]
```diff
@@ -104,20 +106,25 @@ export default {
...
     if (this.onDeselect) {
-      this.onDeselect(this.selectedNodes, this.selectedEdges);
+      this.onDeselect(this.selectedNodes, this.selectedEdges, this.selectedCombos);
     }
 
     graph.emit('nodeselectchange', {
       selectedItems: {
         nodes: [],
         edges: [],
+        combos: [],
```
Comment (Verdagio) [inline packages/pc/src/behavior/brush-select.ts:127]:
```suggestion
        combos: []
```

File: packages/pc/src/behavior/brush-select.ts [hunk 11]
```diff
@@ -19,8 +19,10 @@ export default {
       selectedState: 'selected',
       trigger: DEFAULT_TRIGGER,
       includeEdges: true,
+      includeCombos: false,
       selectedEdges: [],
       selectedNodes: [],
+      selectedCombos: [],
```
Comment (Verdagio) [inline packages/pc/src/behavior/brush-select.ts:25]:
```suggestion
      selectedCombos: []
```

**Review by Verdagio — CHANGES_REQUESTED** [review]
just some minor stuff

File: packages/pc/src/behavior/brush-select.ts (hunk 8 above)
Comment (daniele-pini) [reply packages/pc/src/behavior/brush-select.ts:212]: The existing code already had commas even on the last property, so I preferred to keep the style. If it's a style thing they'd probably as me to re-add those commas in the upstream PR, so let's just preserve the status quo.

File: packages/pc/src/behavior/brush-select.ts (hunk 7 above)
Comment (daniele-pini) [reply packages/pc/src/behavior/brush-select.ts:190]: Hmm... we _may_ also have to do it for edges at some point. Current code "includes" edges when the _source_ node of an edge is included in the selection, which different from what KL does - that goes by midpoint, same check in this `if`.

So yes it makes sense, I'll add it.

**Review by Verdagio — APPROVED** [review]

**Pull request merged**